File: src/test/java/lrusso96/simplebiblio/LibgenTest.java
package lrusso96.simplebiblio;

import lrusso96.simplebiblio.core.Ebook;
import lrusso96.simplebiblio.core.providers.libgen.Field;
import lrusso96.simplebiblio.core.providers.libgen.LibraryGenesis;
import lrusso96.simplebiblio.core.providers.libgen.LibraryGenesisBuilder;
import lrusso96.simplebiblio.core.providers.libgen.Sorting;
import lrusso96.simplebiblio.exceptions.BiblioException;
import org.junit.Test;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;

import static lrusso96.simplebiblio.core.Utils.bytesToReadableSize;
import static org.junit.Assert.*;

public class LibgenTest {

    @Test
    public void simpleSearchTest() throws BiblioException, InterruptedException {
        LibraryGenesis libgen = new LibraryGenesisBuilder().build();
        List<Ebook> ret = libgen.search("Carroll");
        simpleTest(ret);
    }

    @Test
    public void recentTest() throws BiblioException, InterruptedException {
        LibraryGenesis libgen = new LibraryGenesisBuilder().build();
        List<Ebook> ret = libgen.getRecent();
        simpleTest(ret);
    }

    private void simpleTest(List<Ebook> ret) throws InterruptedException {
        assertNotEquals(0, ret.size());
        Ebook book = ret.get(0);
        assertNotNull(book.getAuthor());
        assertNotEquals(0, book.getId());
        assertNotNull(book.getTitle());
        Thread.sleep(2000);
        assertNotEquals(0, book.getDownloads().size());
        assertNotEquals(0, book.getDownloads().get(0).getExtension().length());
        System.out.println(String.format("filesize: %s", bytesToReadableSize(book.getFilesize())));
    }

    @Test
    public void customOptionsTest() throws BiblioException, URISyntaxException {
        LibraryGenesis libgen = new LibraryGenesisBuilder()
                .setMirror(new URI("http://93.174.95.27"))
                .setMaxResultsNumber(10)
                .setSortingField(Field.TITLE)
                .setSortingMode(Sorting.ASCENDING)
                .build();
        List<Ebook> ret = libgen.search("Dante Alighieri");
        assertTrue(ret.size() > 1);
        assertTrue(ret.size() <= 10);
        Ebook b1 = ret.get(0);
        Ebook b2 = ret.get(1);
        assertTrue(b1.getTitle().compareTo(b2.getTitle()) <= 0);
    }
}

File: src/test/java/lrusso96/simplebiblio/LibgenTest.kt
package lrusso96.simplebiblio

import kotlinx.coroutines.runBlocking
import lrusso96.simplebiblio.core.Ebook
import lrusso96.simplebiblio.core.Utils.bytesToReadableSize
import lrusso96.simplebiblio.core.providers.libgen.Field
import lrusso96.simplebiblio.core.providers.libgen.LibraryGenesis
import lrusso96.simplebiblio.core.providers.libgen.Sorting
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.net.URI

class LibgenTest {
    @Test
    fun simpleSearchTest() {
        val libgen = LibraryGenesis.Builder().build()
        val ret = runBlocking { libgen.search("Carroll") }
        simpleTest(ret)
    }

    @Test
    fun recentTest() {
        val libgen = LibraryGenesis.Builder().build()
        val ret = runBlocking { libgen.getRecent() }
        simpleTest(ret)
    }

    private fun simpleTest(ret: List<Ebook>) {
        assertTrue(ret.isNotEmpty())
        val book = ret[0]
        assertNotNull(book.author)
        assertNotEquals(0, book.id)
        assertNotNull(book.title)
        Thread.sleep(2000)
        assertTrue(book.getDownloads().isNotEmpty())
        assertTrue(book.getDownloads()[0].extension!!.isNotBlank())
        println(String.format("filesize: %s", bytesToReadableSize(book.filesize)))
    }

    @Test
    fun customOptionsTest() {
        val libgen = LibraryGenesis.Builder(
                mirror = URI("http://93.174.95.27"),
                maxResults = 10,
                sortingField = Field.TITLE,
                sortingMode = Sorting.ASCENDING)
                .build()
        val ret = runBlocking { libgen.search("Dante Alighieri") }
        assertTrue(ret.size > 1)
        assertTrue(ret.size <= 10)
        val b1 = ret[0]
        val b2 = ret[1]
        assertTrue(b1.title!! <= b2.title!!)
    }
}